<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>ProteinSequence &#8212; cogent3 2022.10.31a1 documentation</title>
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/bootstrap-sphinx.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/jupyter-sphinx.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/thebelab.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sg_gallery.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-bootstrap.5fd3999ee7762ccc51105388f4a9d115.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/thebelab-helper.js"></script>
    <script src="../../../_static/plotly-latest.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.20.0/dist/embed-amd.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="../../../_static/require.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="genetic_code" href="../../genetic_code/genetic_code.html" />
    <link rel="prev" title="RnaSequence" href="cogent3.core.sequence.RnaSequence.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<script type="text/javascript" src="../../../_static/js/jquery-1.12.4.min.js "></script>
<script type="text/javascript" src="../../../_static/js/jquery-fix.js "></script>
<script type="text/javascript" src="../../../_static/bootstrap-3.4.1/js/bootstrap.min.js "></script>
<script type="text/javascript" src="../../../_static/bootstrap-sphinx.js "></script>

  </head><body>

  <div id="navbar" class="navbar navbar-inverse navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../../../index.html">
          Docs</a>
        <span class="navbar-text navbar-version pull-left"><b></b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="../../../install.html">Install</a></li>
                <li><a href="../../../draw/index.html">Gallery</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../../../index.html">Sections <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../app/index.html">The <code class="docutils literal notranslate"><span class="pre">apps</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../cookbook/index.html">Cookbook</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../examples/index.html">Tutorials</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../index.html">API</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">ProteinSequence</a></li>
</ul>
</ul>
</li>
              
            
            
              
                
  <li>
    <a href="cogent3.core.sequence.RnaSequence.html" title="Previous Chapter: RnaSequence"><span class="glyphicon glyphicon-chevron-left visible-sm"></span><span class="hidden-sm hidden-tablet">&laquo; RnaSequence</span>
    </a>
  </li>
  <li>
    <a href="../../genetic_code/genetic_code.html" title="Next Chapter: genetic_code"><span class="glyphicon glyphicon-chevron-right visible-sm"></span><span class="hidden-sm hidden-tablet">genetic_code &raquo;</span>
    </a>
  </li>
              
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../../../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="body col-md-12 content" role="main">
      
  <section id="proteinsequence">
<h1>ProteinSequence<a class="headerlink" href="#proteinsequence" title="Permalink to this headline">¶</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ProteinSequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">info</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">preserve_case</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gaps_allowed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wildcards_allowed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Holds the standard Protein sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>line_wrap</strong></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.annotate_from_gff" title="cogent3.core.sequence.ProteinSequence.annotate_from_gff"><code class="xref py py-obj docutils literal notranslate"><span class="pre">annotate_from_gff</span></code></a>(f[, pre_parsed])</p></td>
<td><p>annotates a Sequence from a gff file where each entry has the same SeqID</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.annotate_matches_to" title="cogent3.core.sequence.ProteinSequence.annotate_matches_to"><code class="xref py py-obj docutils literal notranslate"><span class="pre">annotate_matches_to</span></code></a>(pattern, annot_type, name)</p></td>
<td><p>Adds an annotation at sequence positions matching pattern.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.can_match" title="cogent3.core.sequence.ProteinSequence.can_match"><code class="xref py py-obj docutils literal notranslate"><span class="pre">can_match</span></code></a>(other)</p></td>
<td><p>Returns True if every pos in self could match same pos in other.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.can_mismatch" title="cogent3.core.sequence.ProteinSequence.can_mismatch"><code class="xref py py-obj docutils literal notranslate"><span class="pre">can_mismatch</span></code></a>(other)</p></td>
<td><p>Returns True if any position in self could mismatch with other.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.can_mispair" title="cogent3.core.sequence.ProteinSequence.can_mispair"><code class="xref py py-obj docutils literal notranslate"><span class="pre">can_mispair</span></code></a>(other)</p></td>
<td><p>Returns True if any position in self could mispair with other.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.can_pair" title="cogent3.core.sequence.ProteinSequence.can_pair"><code class="xref py py-obj docutils literal notranslate"><span class="pre">can_pair</span></code></a>(other)</p></td>
<td><p>Returns True if self and other could pair.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.complement" title="cogent3.core.sequence.ProteinSequence.complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">complement</span></code></a>()</p></td>
<td><p>Returns complement of self, using data from MolType.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.copy" title="cogent3.core.sequence.ProteinSequence.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a>()</p></td>
<td><p>returns a copy of self</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.count" title="cogent3.core.sequence.ProteinSequence.count"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count</span></code></a>(item)</p></td>
<td><p>count() delegates to self._seq.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.count_degenerate" title="cogent3.core.sequence.ProteinSequence.count_degenerate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count_degenerate</span></code></a>()</p></td>
<td><p>Counts the degenerate bases in the specified sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.count_gaps" title="cogent3.core.sequence.ProteinSequence.count_gaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count_gaps</span></code></a>()</p></td>
<td><p>Counts the gaps in the specified sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.counts" title="cogent3.core.sequence.ProteinSequence.counts"><code class="xref py py-obj docutils literal notranslate"><span class="pre">counts</span></code></a>([motif_length, include_ambiguity, ...])</p></td>
<td><p>returns dict of counts of motifs</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.degap" title="cogent3.core.sequence.ProteinSequence.degap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">degap</span></code></a>()</p></td>
<td><p>Deletes all gap characters from sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.diff" title="cogent3.core.sequence.ProteinSequence.diff"><code class="xref py py-obj docutils literal notranslate"><span class="pre">diff</span></code></a>(other)</p></td>
<td><p>Returns number of differences between self and other.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.disambiguate" title="cogent3.core.sequence.ProteinSequence.disambiguate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">disambiguate</span></code></a>([method])</p></td>
<td><p>Returns a non-degenerate sequence from a degenerate one.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.distance" title="cogent3.core.sequence.ProteinSequence.distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">distance</span></code></a>(other[, function])</p></td>
<td><p>Returns distance between self and other using function(i,j).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.first_degenerate" title="cogent3.core.sequence.ProteinSequence.first_degenerate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">first_degenerate</span></code></a>()</p></td>
<td><p>Returns the index of first degenerate symbol in sequence, or None.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.first_gap" title="cogent3.core.sequence.ProteinSequence.first_gap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">first_gap</span></code></a>()</p></td>
<td><p>Returns the index of the first gap in the sequence, or None.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.first_invalid" title="cogent3.core.sequence.ProteinSequence.first_invalid"><code class="xref py py-obj docutils literal notranslate"><span class="pre">first_invalid</span></code></a>()</p></td>
<td><p>Returns the index of first invalid symbol in sequence, or None.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.first_non_strict" title="cogent3.core.sequence.ProteinSequence.first_non_strict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">first_non_strict</span></code></a>()</p></td>
<td><p>Returns the index of first non-strict symbol in sequence, or None.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.frac_diff" title="cogent3.core.sequence.ProteinSequence.frac_diff"><code class="xref py py-obj docutils literal notranslate"><span class="pre">frac_diff</span></code></a>(other)</p></td>
<td><p>Returns fraction of positions where self and other differ.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.frac_diff_gaps" title="cogent3.core.sequence.ProteinSequence.frac_diff_gaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">frac_diff_gaps</span></code></a>(other)</p></td>
<td><p>Returns frac.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.frac_diff_non_gaps" title="cogent3.core.sequence.ProteinSequence.frac_diff_non_gaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">frac_diff_non_gaps</span></code></a>(other)</p></td>
<td><p>Returns fraction of non-gap positions where self differs from other.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.frac_same" title="cogent3.core.sequence.ProteinSequence.frac_same"><code class="xref py py-obj docutils literal notranslate"><span class="pre">frac_same</span></code></a>(other)</p></td>
<td><p>Returns fraction of positions where self and other are the same.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.frac_same_gaps" title="cogent3.core.sequence.ProteinSequence.frac_same_gaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">frac_same_gaps</span></code></a>(other)</p></td>
<td><p>Returns fraction of positions where self and other share gap states.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.frac_same_non_gaps" title="cogent3.core.sequence.ProteinSequence.frac_same_non_gaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">frac_same_non_gaps</span></code></a>(other)</p></td>
<td><p>Returns fraction of non-gap positions where self matches other.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.frac_similar" title="cogent3.core.sequence.ProteinSequence.frac_similar"><code class="xref py py-obj docutils literal notranslate"><span class="pre">frac_similar</span></code></a>(other, similar_pairs)</p></td>
<td><p>Returns fraction of positions where self[i] is similar to other[i].</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.gap_indices" title="cogent3.core.sequence.ProteinSequence.gap_indices"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gap_indices</span></code></a>()</p></td>
<td><p>Returns list of indices of all gaps in the sequence, or [].</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.gap_maps" title="cogent3.core.sequence.ProteinSequence.gap_maps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gap_maps</span></code></a>()</p></td>
<td><p>Returns dicts mapping between gapped and ungapped positions.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.gap_vector" title="cogent3.core.sequence.ProteinSequence.gap_vector"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gap_vector</span></code></a>()</p></td>
<td><p>Returns vector of True or False according to which pos are gaps.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.get_annotations_matching" title="cogent3.core.sequence.ProteinSequence.get_annotations_matching"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_annotations_matching</span></code></a>(annotation_type[, ...])</p></td>
<td><p><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.get_by_annotation" title="cogent3.core.sequence.ProteinSequence.get_by_annotation"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_by_annotation</span></code></a>(annotation_type[, name, ...])</p></td>
<td><p>yields the sequence segments corresponding to the specified annotation_type and name one at a time.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.get_drawable" title="cogent3.core.sequence.ProteinSequence.get_drawable"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_drawable</span></code></a>([width, vertical])</p></td>
<td><p>returns Drawable instance</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.get_drawables" title="cogent3.core.sequence.ProteinSequence.get_drawables"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_drawables</span></code></a>()</p></td>
<td><p>returns a dict of drawables, keyed by type</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.get_in_motif_size" title="cogent3.core.sequence.ProteinSequence.get_in_motif_size"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_in_motif_size</span></code></a>([motif_length, log_warnings])</p></td>
<td><p>returns sequence as list of non-overlapping motifs</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.get_kmers" title="cogent3.core.sequence.ProteinSequence.get_kmers"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_kmers</span></code></a>(k)</p></td>
<td><p>return all overlapping k-mers</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.get_name" title="cogent3.core.sequence.ProteinSequence.get_name"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_name</span></code></a>()</p></td>
<td><p>Return the sequence name -- should just use name instead.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.get_type" title="cogent3.core.sequence.ProteinSequence.get_type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_type</span></code></a>()</p></td>
<td><p>Return the sequence type as moltype label.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.gettype" title="cogent3.core.sequence.ProteinSequence.gettype"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gettype</span></code></a>()</p></td>
<td><p>Return the sequence type.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.is_annotated" title="cogent3.core.sequence.ProteinSequence.is_annotated"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_annotated</span></code></a>()</p></td>
<td><p>returns True if sequence has any annotations</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.is_degenerate" title="cogent3.core.sequence.ProteinSequence.is_degenerate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_degenerate</span></code></a>()</p></td>
<td><p>Returns True if sequence contains degenerate characters.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.is_gap" title="cogent3.core.sequence.ProteinSequence.is_gap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_gap</span></code></a>([char])</p></td>
<td><p>Returns True if char is a gap.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.is_gapped" title="cogent3.core.sequence.ProteinSequence.is_gapped"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_gapped</span></code></a>()</p></td>
<td><p>Returns True if sequence contains gaps.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.is_strict" title="cogent3.core.sequence.ProteinSequence.is_strict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_strict</span></code></a>()</p></td>
<td><p>Returns True if sequence contains only monomers.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.is_valid" title="cogent3.core.sequence.ProteinSequence.is_valid"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_valid</span></code></a>()</p></td>
<td><p>Returns True if sequence contains no items absent from alphabet.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.iter_kmers" title="cogent3.core.sequence.ProteinSequence.iter_kmers"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter_kmers</span></code></a>(k)</p></td>
<td><p>generates all overlapping k-mers</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.matrix_distance" title="cogent3.core.sequence.ProteinSequence.matrix_distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">matrix_distance</span></code></a>(other, matrix)</p></td>
<td><p>Returns distance between self and other using a score matrix.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.must_match" title="cogent3.core.sequence.ProteinSequence.must_match"><code class="xref py py-obj docutils literal notranslate"><span class="pre">must_match</span></code></a>(other)</p></td>
<td><p>Returns True if all positions in self must match positions in other.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.must_pair" title="cogent3.core.sequence.ProteinSequence.must_pair"><code class="xref py py-obj docutils literal notranslate"><span class="pre">must_pair</span></code></a>(other)</p></td>
<td><p>Returns True if all positions in self must pair with other.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.mw" title="cogent3.core.sequence.ProteinSequence.mw"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mw</span></code></a>([method, delta])</p></td>
<td><p>Returns the molecular weight of (one strand of) the sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.possibilities" title="cogent3.core.sequence.ProteinSequence.possibilities"><code class="xref py py-obj docutils literal notranslate"><span class="pre">possibilities</span></code></a>()</p></td>
<td><p>Counts number of possible sequences matching the sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.rc" title="cogent3.core.sequence.ProteinSequence.rc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">rc</span></code></a>()</p></td>
<td><p>Returns reverse complement of self w/ data from MolType.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.replace" title="cogent3.core.sequence.ProteinSequence.replace"><code class="xref py py-obj docutils literal notranslate"><span class="pre">replace</span></code></a>(oldchar, newchar)</p></td>
<td><p>return new instance with oldchar replaced by newchar</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.resolveambiguities" title="cogent3.core.sequence.ProteinSequence.resolveambiguities"><code class="xref py py-obj docutils literal notranslate"><span class="pre">resolveambiguities</span></code></a>()</p></td>
<td><p>Returns a list of tuples of strings.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.resolved_ambiguities" title="cogent3.core.sequence.ProteinSequence.resolved_ambiguities"><code class="xref py py-obj docutils literal notranslate"><span class="pre">resolved_ambiguities</span></code></a>()</p></td>
<td><p>Returns a list of tuples of strings.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.shuffle" title="cogent3.core.sequence.ProteinSequence.shuffle"><code class="xref py py-obj docutils literal notranslate"><span class="pre">shuffle</span></code></a>()</p></td>
<td><p>returns a randomized copy of the Sequence object</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.sliding_windows" title="cogent3.core.sequence.ProteinSequence.sliding_windows"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sliding_windows</span></code></a>(window, step[, start, end])</p></td>
<td><p>Generator function that yield new sequence objects of a given length at a given interval.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.strip_bad" title="cogent3.core.sequence.ProteinSequence.strip_bad"><code class="xref py py-obj docutils literal notranslate"><span class="pre">strip_bad</span></code></a>()</p></td>
<td><p>Removes any symbols not in the alphabet.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.strip_bad_and_gaps" title="cogent3.core.sequence.ProteinSequence.strip_bad_and_gaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">strip_bad_and_gaps</span></code></a>()</p></td>
<td><p>Removes any symbols not in the alphabet, and any gaps.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.strip_degenerate" title="cogent3.core.sequence.ProteinSequence.strip_degenerate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">strip_degenerate</span></code></a>()</p></td>
<td><p>Removes degenerate bases by stripping them out of the sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.to_fasta" title="cogent3.core.sequence.ProteinSequence.to_fasta"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_fasta</span></code></a>([make_seqlabel, block_size])</p></td>
<td><p>Return string of self in FASTA format, no trailing newline</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.to_html" title="cogent3.core.sequence.ProteinSequence.to_html"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_html</span></code></a>([wrap, limit, colors, font_size, ...])</p></td>
<td><p>returns html with embedded styles for sequence colouring</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.to_json" title="cogent3.core.sequence.ProteinSequence.to_json"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_json</span></code></a>()</p></td>
<td><p>returns a json formatted string</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.to_moltype" title="cogent3.core.sequence.ProteinSequence.to_moltype"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_moltype</span></code></a>(moltype)</p></td>
<td><p>returns copy of self with moltype seq</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.to_rich_dict" title="cogent3.core.sequence.ProteinSequence.to_rich_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_rich_dict</span></code></a>()</p></td>
<td><p>returns {'name': name, 'seq': sequence, 'moltype': moltype.label}</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.translate" title="cogent3.core.sequence.ProteinSequence.translate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">translate</span></code></a>(*args, **kwargs)</p></td>
<td><p>returns the result of call str.translate</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.with_masked_annotations" title="cogent3.core.sequence.ProteinSequence.with_masked_annotations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">with_masked_annotations</span></code></a>(annot_types[, ...])</p></td>
<td><p>returns a sequence with annot_types regions replaced by mask_char if shadow is False, otherwise all other regions are masked.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.with_termini_unknown" title="cogent3.core.sequence.ProteinSequence.with_termini_unknown"><code class="xref py py-obj docutils literal notranslate"><span class="pre">with_termini_unknown</span></code></a>()</p></td>
<td><p>Returns copy of sequence with terminal gaps remapped as missing.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 75%" />
<col style="width: 25%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>add_annotation</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>add_feature</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>attach_annotations</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>clear_annotations</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>copy_annotations</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>detach_annotations</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>gapped_by_map</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>gapped_by_map_motif_iter</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>gapped_by_map_segment_iter</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>get_region_covering_all</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>parse_out_gaps</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>strand_symmetry</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.add_annotation">
<span class="sig-name descname"><span class="pre">add_annotation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">klass</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kw</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.add_annotation" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.add_feature">
<span class="sig-name descname"><span class="pre">add_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spans</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.add_feature" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.annotate_from_gff">
<span class="sig-name descname"><span class="pre">annotate_from_gff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pre_parsed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.annotate_from_gff" title="Permalink to this definition">¶</a></dt>
<dd><p>annotates a Sequence from a gff file where each entry has the same SeqID</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.annotate_matches_to">
<span class="sig-name descname"><span class="pre">annotate_matches_to</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pattern</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annot_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_multiple</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.annotate_matches_to" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds an annotation at sequence positions matching pattern.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>pattern</strong><span class="classifier">string</span></dt><dd><p>The search string for which annotations are made. IUPAC ambiguities
are converted to regex on sequences with the appropriate MolType.</p>
</dd>
<dt><strong>annot_type</strong><span class="classifier">string</span></dt><dd><p>The type of the annotation (e.g. “domain”).</p>
</dd>
<dt><strong>name</strong><span class="classifier">string</span></dt><dd><p>The name of the annotation.</p>
</dd>
<dt><strong>allow_multiple</strong><span class="classifier">boolean</span></dt><dd><p>If True, allows multiple occurrences of the input pattern. Otherwise
only the first match is used.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>Returns a list of Annotation instances.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.annotations">
<span class="sig-name descname"><span class="pre">annotations</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">()</span></em><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.annotations" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.attach_annotations">
<span class="sig-name descname"><span class="pre">attach_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">annots</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.attach_annotations" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.can_match">
<span class="sig-name descname"><span class="pre">can_match</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.can_match" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if every pos in self could match same pos in other.</p>
<p>Truncates at length of shorter sequence.
gaps are only allowed to match other gaps.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.can_mismatch">
<span class="sig-name descname"><span class="pre">can_mismatch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.can_mismatch" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if any position in self could mismatch with other.</p>
<p>Truncates at length of shorter sequence.
gaps are always counted as matches.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.can_mispair">
<span class="sig-name descname"><span class="pre">can_mispair</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.can_mispair" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if any position in self could mispair with other.</p>
<p>Pairing occurs in reverse order, i.e. last position of other with
first position of self, etc.</p>
<p>Truncates at length of shorter sequence.
gaps are always counted as possible mispairs, as are weak pairs like GU.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.can_pair">
<span class="sig-name descname"><span class="pre">can_pair</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.can_pair" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if self and other could pair.</p>
<p>Pairing occurs in reverse order, i.e. last position of other with
first position of self, etc.</p>
<p>Truncates at length of shorter sequence.
gaps are only allowed to pair with other gaps, and are counted as ‘weak’
(same category as GU and degenerate pairs).</p>
<p>NOTE: second must be able to be reverse</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.clear_annotations">
<span class="sig-name descname"><span class="pre">clear_annotations</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.clear_annotations" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.complement">
<span class="sig-name descname"><span class="pre">complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns complement of self, using data from MolType.</p>
<p>Always tries to return same type as item: if item looks like a dict,
will return list of keys.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a copy of self</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.copy_annotations">
<span class="sig-name descname"><span class="pre">copy_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.copy_annotations" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.count">
<span class="sig-name descname"><span class="pre">count</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">item</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.count" title="Permalink to this definition">¶</a></dt>
<dd><p>count() delegates to self._seq.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.count_degenerate">
<span class="sig-name descname"><span class="pre">count_degenerate</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.count_degenerate" title="Permalink to this definition">¶</a></dt>
<dd><p>Counts the degenerate bases in the specified sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.count_gaps">
<span class="sig-name descname"><span class="pre">count_gaps</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.count_gaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Counts the gaps in the specified sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.counts">
<span class="sig-name descname"><span class="pre">counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_ambiguity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_unobserved</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.counts" title="Permalink to this definition">¶</a></dt>
<dd><p>returns dict of counts of motifs</p>
<p>only non-overlapping motifs are counted.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>motif_length</strong></dt><dd><p>number of elements per character.</p>
</dd>
<dt><strong>include_ambiguity</strong></dt><dd><p>if True, motifs containing ambiguous characters
from the seq moltype are included. No expansion of those is attempted.</p>
</dd>
<dt><strong>allow_gaps</strong></dt><dd><p>if True, motifs containing a gap character are included.</p>
</dd>
<dt><strong>exclude_unobserved</strong></dt><dd><p>if True, unobserved motif combinations are excluded.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.degap">
<span class="sig-name descname"><span class="pre">degap</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.degap" title="Permalink to this definition">¶</a></dt>
<dd><p>Deletes all gap characters from sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.detach_annotations">
<span class="sig-name descname"><span class="pre">detach_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">annots</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.detach_annotations" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.diff">
<span class="sig-name descname"><span class="pre">diff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.diff" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns number of differences between self and other.</p>
<p>NOTE: truncates at the length of the shorter sequence. Case-sensitive.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.disambiguate">
<span class="sig-name descname"><span class="pre">disambiguate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'strip'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.disambiguate" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a non-degenerate sequence from a degenerate one.</p>
<p>method can be ‘strip’ (deletes any characters not in monomers or gaps)
or ‘random’(assigns the possibilities at random, using equal
frequencies).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.distance">
<span class="sig-name descname"><span class="pre">distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">function</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns distance between self and other using function(i,j).</p>
<p>other must be a sequence.</p>
<p>function should be a function that takes two items and returns a
number. To turn a 2D matrix into a function, use
cogent3.util.miscs.DistanceFromMatrix(matrix).</p>
<p>NOTE: Truncates at the length of the shorter sequence.</p>
<p>Note that the function acts on two _elements_ of the sequences, not
the two sequences themselves (i.e. the behavior will be the same for
every position in the sequences, such as identity scoring or a function
derived from a distance matrix as suggested above). One limitation of
this approach is that the distance function cannot use properties of
the sequences themselves: for example, it cannot use the lengths of the
sequences to normalize the scores as percent similarities or percent
differences.</p>
<p>If you want functions that act on the two sequences themselves, there
is no particular advantage in making these functions methods of the
first sequences by passing them in as parameters like the function
in this method. It makes more sense to use them as standalone functions.
The factory function cogent3.util.transform.for_seq is useful for
converting per-element functions into per-sequence functions, since it
takes as parameters a per-element scoring function, a score aggregation
function, and a normalization function (which itself takes the two
sequences as parameters), returning a single function that combines
these functions and that acts on two complete sequences.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.first_degenerate">
<span class="sig-name descname"><span class="pre">first_degenerate</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.first_degenerate" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the index of first degenerate symbol in sequence, or None.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.first_gap">
<span class="sig-name descname"><span class="pre">first_gap</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.first_gap" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the index of the first gap in the sequence, or None.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.first_invalid">
<span class="sig-name descname"><span class="pre">first_invalid</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.first_invalid" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the index of first invalid symbol in sequence, or None.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.first_non_strict">
<span class="sig-name descname"><span class="pre">first_non_strict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.first_non_strict" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the index of first non-strict symbol in sequence, or None.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.frac_diff">
<span class="sig-name descname"><span class="pre">frac_diff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.frac_diff" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns fraction of positions where self and other differ.</p>
<p>Truncates at length of shorter sequence.
Note that frac_same and frac_diff are both 0 if one sequence is empty.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.frac_diff_gaps">
<span class="sig-name descname"><span class="pre">frac_diff_gaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.frac_diff_gaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns frac. of positions where self and other’s gap states differ.</p>
<p>In other words, if self and other are both all gaps, or both all
non-gaps, or both have gaps in the same places, frac_diff_gaps will
return 0.0. If self is all gaps and other has no gaps, frac_diff_gaps
will return 1.0.</p>
<p>Returns 0 if one sequence is empty.</p>
<p>Uses self’s gap characters for both sequences.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.frac_diff_non_gaps">
<span class="sig-name descname"><span class="pre">frac_diff_non_gaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.frac_diff_non_gaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns fraction of non-gap positions where self differs from other.</p>
<p>Doesn’t count any position where self or other has a gap.
Truncates at the length of the shorter sequence.</p>
<p>Returns 0 if one sequence is empty. Note that this means that
frac_diff_non_gaps is _not_ the same as 1 - frac_same_non_gaps, since both
return 0 if one sequence is empty.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.frac_same">
<span class="sig-name descname"><span class="pre">frac_same</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.frac_same" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns fraction of positions where self and other are the same.</p>
<p>Truncates at length of shorter sequence.
Note that frac_same and frac_diff are both 0 if one sequence is empty.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.frac_same_gaps">
<span class="sig-name descname"><span class="pre">frac_same_gaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.frac_same_gaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns fraction of positions where self and other share gap states.</p>
<p>In other words, if self and other are both all gaps, or both all
non-gaps, or both have gaps in the same places, frac_same_gaps will
return 1.0. If self is all gaps and other has no gaps, frac_same_gaps
will return 0.0. Returns 0 if one sequence is empty.</p>
<p>Uses self’s gap characters for both sequences.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.frac_same_non_gaps">
<span class="sig-name descname"><span class="pre">frac_same_non_gaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.frac_same_non_gaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns fraction of non-gap positions where self matches other.</p>
<p>Doesn’t count any position where self or other has a gap.
Truncates at the length of the shorter sequence.</p>
<p>Returns 0 if one sequence is empty.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.frac_similar">
<span class="sig-name descname"><span class="pre">frac_similar</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">similar_pairs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.frac_similar" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns fraction of positions where self[i] is similar to other[i].</p>
<p>similar_pairs must be a dict such that d[(i,j)] exists if i and j are
to be counted as similar. Use PairsFromGroups in cogent3.util.misc to
construct such a dict from a list of lists of similar residues.</p>
<p>Truncates at the length of the shorter sequence.</p>
<p>Note: current implementation re-creates the distance function each
time, so may be expensive compared to creating the distance function
using for_seq separately.</p>
<p>Returns 0 if one sequence is empty.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.gap_indices">
<span class="sig-name descname"><span class="pre">gap_indices</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.gap_indices" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns list of indices of all gaps in the sequence, or [].</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.gap_maps">
<span class="sig-name descname"><span class="pre">gap_maps</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.gap_maps" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns dicts mapping between gapped and ungapped positions.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.gap_vector">
<span class="sig-name descname"><span class="pre">gap_vector</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.gap_vector" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns vector of True or False according to which pos are gaps.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.gapped_by_map">
<span class="sig-name descname"><span class="pre">gapped_by_map</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">map</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recode_gaps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.gapped_by_map" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.gapped_by_map_motif_iter">
<span class="sig-name descname"><span class="pre">gapped_by_map_motif_iter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">map</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.gapped_by_map_motif_iter" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.gapped_by_map_segment_iter">
<span class="sig-name descname"><span class="pre">gapped_by_map_segment_iter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">map</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_gaps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recode_gaps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.gapped_by_map_segment_iter" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.get_annotations_matching">
<span class="sig-name descname"><span class="pre">get_annotations_matching</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">annotation_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extend_query</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.get_annotations_matching" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>annotation_type</strong><span class="classifier">string</span></dt><dd><p>name of the annotation type. Wild-cards allowed.</p>
</dd>
<dt><strong>name</strong><span class="classifier">string</span></dt><dd><p>name of the instance. Wild-cards allowed.</p>
</dd>
<dt><strong>extend_query</strong><span class="classifier">boolean</span></dt><dd><p>queries sub-annotations if True</p>
</dd>
<dt><strong>Returns</strong></dt><dd></dd>
<dt><strong>——-</strong></dt><dd></dd>
<dt><strong>list of AnnotatableFeatures</strong></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.get_by_annotation">
<span class="sig-name descname"><span class="pre">get_by_annotation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">annotation_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_partial</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.get_by_annotation" title="Permalink to this definition">¶</a></dt>
<dd><p>yields the sequence segments corresponding to the specified
annotation_type and name one at a time.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ignore_partial</strong></dt><dd><p>if True, annotations that extend beyond the
current sequence are ignored.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.get_drawable">
<span class="sig-name descname"><span class="pre">get_drawable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">600</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vertical</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.get_drawable" title="Permalink to this definition">¶</a></dt>
<dd><p>returns Drawable instance</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.get_drawables">
<span class="sig-name descname"><span class="pre">get_drawables</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.get_drawables" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a dict of drawables, keyed by type</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.get_in_motif_size">
<span class="sig-name descname"><span class="pre">get_in_motif_size</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_warnings</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.get_in_motif_size" title="Permalink to this definition">¶</a></dt>
<dd><p>returns sequence as list of non-overlapping motifs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>motif_length</strong></dt><dd><p>length of the motifs</p>
</dd>
<dt><strong>log_warnings</strong></dt><dd><p>whether to notify of an incomplete terminal motif</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.get_kmers">
<span class="sig-name descname"><span class="pre">get_kmers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.get_kmers" title="Permalink to this definition">¶</a></dt>
<dd><p>return all overlapping k-mers</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.get_name">
<span class="sig-name descname"><span class="pre">get_name</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.get_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the sequence name – should just use name instead.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.get_region_covering_all">
<span class="sig-name descname"><span class="pre">get_region_covering_all</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">annotations</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">feature_class</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extend_query</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.get_region_covering_all" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.get_type">
<span class="sig-name descname"><span class="pre">get_type</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.get_type" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the sequence type as moltype label.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.gettype">
<span class="sig-name descname"><span class="pre">gettype</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.gettype" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the sequence type.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.is_annotated">
<span class="sig-name descname"><span class="pre">is_annotated</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.is_annotated" title="Permalink to this definition">¶</a></dt>
<dd><p>returns True if sequence has any annotations</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.is_degenerate">
<span class="sig-name descname"><span class="pre">is_degenerate</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.is_degenerate" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if sequence contains degenerate characters.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.is_gap">
<span class="sig-name descname"><span class="pre">is_gap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">char</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.is_gap" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if char is a gap.</p>
<p>If char is not supplied, tests whether self is gaps only.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.is_gapped">
<span class="sig-name descname"><span class="pre">is_gapped</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.is_gapped" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if sequence contains gaps.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.is_strict">
<span class="sig-name descname"><span class="pre">is_strict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.is_strict" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if sequence contains only monomers.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.is_valid">
<span class="sig-name descname"><span class="pre">is_valid</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.is_valid" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if sequence contains no items absent from alphabet.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.iter_kmers">
<span class="sig-name descname"><span class="pre">iter_kmers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Generator</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">None</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">None</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.iter_kmers" title="Permalink to this definition">¶</a></dt>
<dd><p>generates all overlapping k-mers</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.line_wrap">
<span class="sig-name descname"><span class="pre">line_wrap</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.line_wrap" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.matrix_distance">
<span class="sig-name descname"><span class="pre">matrix_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">matrix</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.matrix_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns distance between self and other using a score matrix.</p>
<p>WARNING: the matrix must explicitly contain scores for the case where
a position is the same in self and other (e.g. for a distance matrix,
an identity between U and U might have a score of 0). The reason the
scores for the ‘diagonals’ need to be passed explicitly is that for
some kinds of distance matrices, e.g. log-odds matrices, the ‘diagonal’
scores differ from each other. If these elements are missing, this
function will raise a KeyError at the first position that the two
sequences are identical.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.moltype">
<span class="sig-name descname"><span class="pre">moltype</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">MolType(('A',</span> <span class="pre">'C',</span> <span class="pre">'D',</span> <span class="pre">'E',</span> <span class="pre">'F',</span> <span class="pre">'G',</span> <span class="pre">'H',</span> <span class="pre">'I',</span> <span class="pre">'K',</span> <span class="pre">'L',</span> <span class="pre">'M',</span> <span class="pre">'N',</span> <span class="pre">'P',</span> <span class="pre">'Q',</span> <span class="pre">'R',</span> <span class="pre">'S',</span> <span class="pre">'T',</span> <span class="pre">'U',</span> <span class="pre">'V',</span> <span class="pre">'W',</span> <span class="pre">'Y'))</span></em><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.moltype" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.must_match">
<span class="sig-name descname"><span class="pre">must_match</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.must_match" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if all positions in self must match positions in other.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.must_pair">
<span class="sig-name descname"><span class="pre">must_pair</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.must_pair" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if all positions in self must pair with other.</p>
<p>Pairing occurs in reverse order, i.e. last position of other with
first position of self, etc.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.mw">
<span class="sig-name descname"><span class="pre">mw</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'random'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.mw" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the molecular weight of (one strand of) the sequence.</p>
<p>If the sequence is ambiguous, uses method (random or strip) to
disambiguate the sequence.</p>
<p>If delta is passed in, adds delta per strand (default is None, which
uses the alphabet default. Typically, this adds 18 Da for terminal
water. However, note that the default nucleic acid weight assumes
5’ monophosphate and 3’ OH: pass in delta=18.0 if you want 5’ OH as
well.</p>
<p>Note that this method only calculates the MW of the coding strand. If
you want the MW of the reverse strand, add self.rc().mw(). DO NOT
just multiply the MW by 2: the results may not be accurate due to
strand bias, e.g. in mitochondrial genomes.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.parse_out_gaps">
<span class="sig-name descname"><span class="pre">parse_out_gaps</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.parse_out_gaps" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.possibilities">
<span class="sig-name descname"><span class="pre">possibilities</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.possibilities" title="Permalink to this definition">¶</a></dt>
<dd><p>Counts number of possible sequences matching the sequence.</p>
<p>Uses self.degenerates to decide how many possibilities there are at
each position in the sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.rc" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns reverse complement of self w/ data from MolType.</p>
<p>Always returns same type self.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.replace">
<span class="sig-name descname"><span class="pre">replace</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">oldchar</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">newchar</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.replace" title="Permalink to this definition">¶</a></dt>
<dd><p>return new instance with oldchar replaced by newchar</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.resolveambiguities">
<span class="sig-name descname"><span class="pre">resolveambiguities</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.resolveambiguities" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of tuples of strings.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.resolved_ambiguities">
<span class="sig-name descname"><span class="pre">resolved_ambiguities</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.resolved_ambiguities" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of tuples of strings.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.shuffle">
<span class="sig-name descname"><span class="pre">shuffle</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.shuffle" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a randomized copy of the Sequence object</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.sliding_windows">
<span class="sig-name descname"><span class="pre">sliding_windows</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">window</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">step</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.sliding_windows" title="Permalink to this definition">¶</a></dt>
<dd><p>Generator function that yield new sequence objects
of a given length at a given interval.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>window</strong></dt><dd><p>The length of the returned sequence</p>
</dd>
<dt><strong>step</strong></dt><dd><p>The interval between the start of the returned
sequence objects</p>
</dd>
<dt><strong>start</strong></dt><dd><p>first window start position</p>
</dd>
<dt><strong>end</strong></dt><dd><p>last window start position</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.strand_symmetry">
<span class="sig-name descname"><span class="pre">strand_symmetry</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.strand_symmetry" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.strip_bad">
<span class="sig-name descname"><span class="pre">strip_bad</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.strip_bad" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes any symbols not in the alphabet.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.strip_bad_and_gaps">
<span class="sig-name descname"><span class="pre">strip_bad_and_gaps</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.strip_bad_and_gaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes any symbols not in the alphabet, and any gaps.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.strip_degenerate">
<span class="sig-name descname"><span class="pre">strip_degenerate</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.strip_degenerate" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes degenerate bases by stripping them out of the sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.to_fasta">
<span class="sig-name descname"><span class="pre">to_fasta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">make_seqlabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">60</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.to_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Return string of self in FASTA format, no trailing newline</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>make_seqlabel</strong></dt><dd><p>callback function that takes the seq object and
returns a label str</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.to_html">
<span class="sig-name descname"><span class="pre">to_html</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">60</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">font_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">12</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">font_family</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Lucida</span> <span class="pre">Console'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.to_html" title="Permalink to this definition">¶</a></dt>
<dd><p>returns html with embedded styles for sequence colouring</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl>
<dt><strong>wrap</strong></dt><dd><p>maximum number of printed bases, defaults to
alignment length</p>
</dd>
<dt><strong>limit</strong></dt><dd><p>truncate alignment to this length</p>
</dd>
<dt><strong>colors</strong></dt><dd><p>{character
moltype.</p>
</dd>
<dt><strong>font_size</strong></dt><dd><p>in points. Affects labels and sequence and line spacing
(proportional to value)</p>
</dd>
<dt><strong>font_family</strong></dt><dd><p>string denoting font family</p>
</dd>
<dt><strong>To display in jupyter notebook:</strong></dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">IPython.core.display</span> <span class="kn">import</span> <span class="n">HTML</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">HTML</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">to_html</span><span class="p">())</span>
</pre></div>
</div>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.to_json">
<span class="sig-name descname"><span class="pre">to_json</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.to_json" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a json formatted string</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.to_moltype">
<span class="sig-name descname"><span class="pre">to_moltype</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">moltype</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.to_moltype" title="Permalink to this definition">¶</a></dt>
<dd><p>returns copy of self with moltype seq</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>moltype</strong><span class="classifier">str</span></dt><dd><p>molecular type</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.to_rich_dict">
<span class="sig-name descname"><span class="pre">to_rich_dict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.to_rich_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>returns {‘name’: name, ‘seq’: sequence, ‘moltype’: moltype.label}</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.translate">
<span class="sig-name descname"><span class="pre">translate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.translate" title="Permalink to this definition">¶</a></dt>
<dd><p>returns the result of call str.translate</p>
<p class="rubric">Notes</p>
<p>This is a string method, nothing to do with translating into a
protein sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.with_masked_annotations">
<span class="sig-name descname"><span class="pre">with_masked_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">annot_types</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_char</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shadow</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extend_query</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.with_masked_annotations" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a sequence with annot_types regions replaced by mask_char
if shadow is False, otherwise all other regions are masked.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>annot_types</strong></dt><dd><p>annotation type(s)</p>
</dd>
<dt><strong>mask_char</strong></dt><dd><p>must be a character valid for the seq MolType. The
default value is the most ambiguous character, eg. ‘?’ for DNA</p>
</dd>
<dt><strong>shadow</strong></dt><dd><p>whether to mask the annotated regions, or everything but
the annotated regions</p>
</dd>
<dt><strong>extend_query</strong><span class="classifier">boolean</span></dt><dd><p>queries sub-annotations if True</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.with_termini_unknown">
<span class="sig-name descname"><span class="pre">with_termini_unknown</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.with_termini_unknown" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns copy of sequence with terminal gaps remapped as missing.</p>
</dd></dl>

</dd></dl>

</section>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2020-2023, cogent3 Team.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 4.4.0.<br/>
    </p>
  </div>
</footer>
  </body>
</html>